File: src/room.rs
use location::Location;
use inventory_item::InventoryItem;
use npc::NPC;

// We want to be able to debug Room and `==` it.
#[derive(Debug, Eq, PartialEq)]
pub struct Room {
    pub location: Location, // it has a Location
    pub name: String, // it has a name
    pub description: String, // it has a description
    pub items: Vec<InventoryItem>, // it has items/inventory (may be empty)
    pub npc: NPC, // it may have an NPC / non-player character
}

impl Room {
    // As in Ruby, we make Room::new take an x, y and description,
    // and construct a Location using the x and y.
    pub fn new(x: u64,
           y: u64,
           name: String,
           description: String,
           items: Vec<InventoryItem>,
           npc: NPC)
           -> Room {
        // returns a Room
        Room {
            // Construct a Room
            location: Location::new(x, y), // construct a Location
            name: name, // save the name
            description: description, // save the description
            items: items, // save the inventory
            npc: npc, // save the NPC
        }
    }

    pub fn items(&self) -> &[InventoryItem] {
        &self.items[..]
    }

    pub fn items_mut(&mut self) -> &mut Vec<InventoryItem> {
        &mut self.items
    }

    pub fn npc(&self) -> &NPC {
        &self.npc
    }

    pub fn npc_mut(&mut self) -> &mut NPC {
        &mut self.npc
    }
}

File: lib/room.rb
# A Room is a simple object that has a `location`,
# (which is a Location object) and a description
# (which is a string).
class Room < Struct.new(:location, :name, :description, :items, :npc)
  # `items` is the Room's inventory, which is empty
  # unless populated on Room initialization
  @items = []
  # The Room is constructed with an x and y, which
  # are passed to Location.new, so they should be
  # numbers.
  def self.new(x, y, name, description, items, npc)
    super Location.new(x, y), name, description, items, npc
  end

  # to_s prints out the x and y coordinates and
  # the name.
  def to_s
    "#{location.x}, #{location.y}, #{name}"
  end

  def has_items
    self.items.any?
  end

  # check to see if this Room has this item
  def has_item(item)
    item_names = []
    self.items.compact.each { |item| item_names << item.name }
    item_names.include? item
  end

  def remove_one(item)
    room_item = self.items.find { |i| i == item }
    # can't remove an item that's not there
    return if !room_item
    index = self.items.index(room_item)
    self.items.delete_at(index)
  end

  # puts-able list of items in Room
  def item_list
    if self.items.compact.count == 1
      self.items.compact.first.name
    else
      item_names = []
      self.items.compact.each { |item| item_names << item.name }
      item_names.join(", ")
    end
  end
end
